skipper.go

```diff
@@ -1198,9 +1373,20 @@ func run(o Options, sig chan os.Signal, idleConnsCH chan struct{}) error {
 		},
 		SignalFirstLoad: o.WaitFirstRouteLoad,
 	}
+
 	if o.DefaultFilters != nil {
-		ro.PreProcessors = []routing.PreProcessor{o.DefaultFilters}
+		ro.PreProcessors = append(ro.PreProcessors, o.DefaultFilters)
+	}
+
+	if o.EnableOAuth2GrantFlow /* explicitly enable grant flow when callback route was not disabled */ {
+		grantPrep, err := oauthConfig.NewGrantPreprocessor()
+		if err != nil {
+			log.Errorf("Error while initializing oauth grant preprocessor: %v.", err)
+		}
+
+		ro.PreProcessors = append(ro.PreProcessors, grantPrep)
```

Comment:
With `ro.PreProcessors = append(ro.PreProcessors, o.DefaultFilters)` change there is not need to have second `if o.EnableOAuth2GrantFlow` - grant preprocessor could be appended in the filter construction block to keep all grant init in one place.

Reply:
Is it safe to move this to the filter construction block? It would then result in adding the grant preprocessor to the `ro.PreProcessors` _before_ `o.DefaultFilters`, which currently is right above:
```go
	if o.DefaultFilters != nil {
		ro.PreProcessors = append(ro.PreProcessors, o.DefaultFilters)
	}
```
Wouldn't it affect the order in which the default filters are applied?

Reply:
As far as I can tell default filters add filters to routes and grant preprocessor adds a route. So if grant preprocessor is in the list _before_ default filters preprocessor they would be added to grant route as well which I guess is desirable. 
https://github.com/zalando/skipper/blob/11e4b519fd7b85b2bf0ea39c1bcac504a5544303/routing/datasource.go#L535-L537

Reply:
Do we create an automated call back route for the redirect?
I am not sure if implicit routes like this were a good choice.

Reply:
> Do we create an automated call back route for the redirect?

Yes that's what it looks like, and I completely missed that in the original PR, assuming that the callback should be explicit. Either the implicit route should be removed, or the documentation needs to be updated to reflect the fact that no explicit callback route is necessary. I'll defer to your judgement here on the best approach.

Reply:
I am checking this now, I just need to recollect the ideas behind my original approach :smile: 

The concern with the order of the DefaultFilters and the grant preprocessor seems to be valid.

@szuecs we should not have the grant filter on http routes, only on https routes. So we should generate the redirect route for the http routes, either. Do I misunderstand your concern?

Reply:
> So if grant preprocessor is in the list before default filters preprocessor they would be added to grant route as well which I guess is desirable.

I think we can argue both ways. The callback route is an implicitly generated route, not defined by the users/operator, so it may make sense that the default filters are not there. Imagine, when somebody uses a tokeninfo filter as a default filter, because in their setup it makes sense. But I see how we can argue the other way around just as well, e.g. because default filters should be just that: "default".

My suggestion is that to start, let's _not_ add the default filters to the callback route, and see in practice how far that brings us. If it turns out to be wrong, it's easier to fix that in a later pass.

WDYT?

Reply:
Discussed in chat:
   > One problem is that it limits the number of providers for one skipper deployment, because callback route is a single route with no    dispatch to another provider. We have also hard coded configuration right now to have only one provider possible anyways, so maybe this is fine right now. I think in the future we should think about how we can create less friction (check oidc filters are very different and build to use multiple providers), but have increased usability (this one is much better than oidc filters).

I would add default filters to the route as is, because you likely want to have lifo() for example set there, too. I don't think tokeninfo filters will be used in default filters.